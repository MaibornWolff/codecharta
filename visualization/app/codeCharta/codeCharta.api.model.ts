import { AttributeTypes, AttributeTypeValue, CodeMapNode, Edge, MarkedPackage, RecursivePartial, Settings } from "./codeCharta.model"

export interface ExportCCFile {
	projectName: string
	apiVersion: string
	nodes: CodeMapNode[]
	attributeTypes?: AttributeTypes | OldAttributeTypes
	edges?: Edge[]
	markedPackages?: MarkedPackage[]
	blacklist?: ExportBlacklistItem[]
}

export interface ExportBlacklistItem {
	path: string
	type: ExportBlacklistType
}

export enum ExportBlacklistType {
	hide = "hide",
	exclude = "exclude"
}

export enum APIVersions {
	ZERO_POINT_ONE = "0.1",
	ONE_POINT_ZERO = "1.0",
	ONE_POINT_ONE = "1.1"
}

export interface ExportScenario {
	name: string
	settings: RecursivePartial<Settings>
}

export interface OldAttributeTypes {
	nodes?: [{ [key: string]: AttributeTypeValue }]
	edges?: [{ [key: string]: AttributeTypeValue }]
}
