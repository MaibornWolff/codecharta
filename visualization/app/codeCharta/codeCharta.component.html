<tool-bar-component></tool-bar-component>

<file-extension-bar-component></file-extension-bar-component>
<div class="ribbon-bar-wrapper">
	<ribbon-bar-component class="ribbon-bar-component"></ribbon-bar-component>
</div>

<code-map-component></code-map-component>

<legend-panel-component></legend-panel-component>

<loading-gif-file-component></loading-gif-file-component>

<div id="mw-logo">
	<div class="logo">
		<a href="https://www.maibornwolff.de" target="_blank"><img src="assets/logo.png" alt="MaibornWolff"/></a>
		<h2>
			<span>CodeCharta {{ $ctrl._viewModel.version }}</span>
		</h2>
	</div>
</div>

<node-context-menu-component></node-context-menu-component>
