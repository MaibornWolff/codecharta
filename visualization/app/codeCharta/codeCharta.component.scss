@import "../mixins.scss";

code-charta-component {
	#mw-logo {
		.logo {
			width: 160px;
			height: 100px;
			position: absolute;
			left: 0;
			bottom: 30px;
			z-index: 29;
		}

		a,
		img {
			width: 5em;
			display: block;
			margin: auto !important;
		}

		h2 {
			text-transform: uppercase;
			color: #5a585a;
			font-weight: 500;
			font-size: 9pt;
			text-align: center;
			margin: 0;
			z-index: -1;
		}
	}

	#codeMap {
		position: fixed;
		z-index: 0;
		top: 98px;
		left: 0;
		width: 100%;
		height: 100%;
	}

	.cc-shadow {
		box-shadow: 0 1px 5px 0 rgba(0, 0, 0, 0.2) !important;
	}

	.ribbon-bar-wrapper {
		height: 500px;
		position: fixed;
		z-index: 10;
		white-space: nowrap;
		pointer-events: none;
	}

	attribute-side-bar-component .side-bar-container,
	legend-panel-component .block-wrapper,
	legend-panel-component .panel-button,
	view-cube-component {
		transition: right 0.25s ease;
	}

	.angular-green {
		color: #00bfa5;
	}

	.small-action-button {
		display: inline-block;
		height: 24px;
		width: 24px;
		vertical-align: top;

		.md-fab {
			min-height: 0;
			width: 100%;
			height: 100%;
			margin: 0;

			i {
				position: absolute;
				top: 5px;
				left: 6px;
			}
		}

		&.transparent .md-fab {
			background-color: transparent !important;
			box-shadow: none;

			&:hover {
				background-color: #ddd !important;
			}
		}
	}

	.md-button.round-button {
		margin-right: 0 !important;
		float: right;
		min-width: 36px;
		width: 36px;
		border-radius: 50%;
		margin: 0;
	}

	.action-buttons {
		position: fixed;
		bottom: 2em;
		right: 2em;
		z-index: 60;

		.long-button {
			width: inherit;
			padding-left: 10px;
			padding-right: 10px;
			border-radius: 100px;
		}

		> * {
			margin-left: 4px;
			margin-right: 4px;
			z-index: 60;
		}
	}

	.state-selector-button {
		height: 22px;
		vertical-align: bottom;
		width: 30px;
		min-height: 0;
		min-width: 0;
		margin: 0;
		padding: 0;
		font-size: 9pt;
		text-transform: none;
		background-color: rgba(230, 230, 230, 0.96);

		&:hover {
			background-color: rgb(200, 200, 200) !important;
		}

		i {
			vertical-align: middle;
			height: 26px;
		}

		&.left {
			border-radius: 5px 0 0 5px;
		}

		&.middle {
			border-radius: 0;
			border-left: 1px solid rgba(255, 255, 255, 1);
			border-right: 1px solid rgba(255, 255, 255, 1);
		}

		&.right {
			border-radius: 0 5px 5px 0;
		}

		&.current {
			background-color: rgb(0, 150, 136);
			color: #ffffff;
		}

		&.current:hover {
			background-color: rgb(0, 124, 112) !important;
		}
	}
}
