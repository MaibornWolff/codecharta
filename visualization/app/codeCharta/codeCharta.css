*, *:before, *:after {
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
}

html, body {
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    width: 100%;
    height: 100%;
    margin: 0;
    padding: 0;
    overflow: hidden;
    font-size: 1em;
    cursor: default;
    color: #5A585A;
    font-family: Roboto, sans-serif;
    font-weight: 300;
}

#codeMap {
    position: absolute;
    z-index: 0;
    top: 50px;
    left: 0;
    width: 100%;
    height: 100%;
}

#hud {
    position: absolute;
    z-index: 10;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    max-width: none !important;
    pointer-events: none;
    padding-left:10px;
    padding-right:10px;
}

revision-chooser-component, legend-panel-component, file-chooser-panel-directive, settings-panel-component, .mwlink {
    pointer-events: auto;
}

#title {
    text-transform: uppercase;
    font-size: 0.9em;
    line-height: 1.4em;
    float: right;
    margin: 0.5em 1.6em 0 0;
    color: #5A585A;
    font-family: Roboto, sans-serif;
    font-weight: 300;
}

#title:first-child {
    font-size: 1.82em;
}

#title:last-child {
    font-size: 1.5em;
}

.collapsible > li {
    background-color: #ffffff !important;
}

.col-backdrop {
    background-color: rgba(0,0,0,0.1);
}

.bottom-row {
    position: fixed;
    bottom: 0;
    width:100%;
}

.mwlink > img{
    width: 100%;
    max-width: 4em;
    float:right;
}

.chip {
    background-color: #efefef;
}

.row, .col {
    min-height: 0px !important;
}


/*

#actionButtons {
    position: fixed;
    bottom: 2em;
    right: 2em;
    z-index: 1000;
}

#settingsPanel {
    position: fixed;
    right: 0;
    width: 0px;
    height: 100%;
    background: #fff;
    z-index: 1000;
    box-shadow: 0 0 10px #000;
}

#detailWindow {
    position: fixed;
    z-index: 999;
    bottom: 0.5em;
    left: 0.5em;
    color:#5A585A;
    font-size: 0.8em;
}

#legendButton {
    position: absolute;
    top: 11.5em;
    left: -2.9em;
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}

#legendPanel {
    position: fixed;
    top: 9em;
    left: -500px;
}

#revisionButton {
    position: absolute;
    top: 21em;
    left: -3.5em;
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}

#mapButton {
    position: absolute;
    top: 21em;
    left: -2.3em;
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}

#revisionChooser {
    position: fixed;
    top: 18.5em;
    left: -500px;
}

.rzslider .rz-pointer, .rzslider .rz-selection {
    background: #26a69a !important;
}

.rzslider .rz-bar {
    background: rgb(196,196,196);
}*/