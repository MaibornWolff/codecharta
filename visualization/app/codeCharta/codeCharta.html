<div id="hud" class="container">

    <div class="row">

        <div class="col l3 m6 s12 col-backdrop">
            <legend-panel-component></legend-panel-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <revision-chooser-component></revision-chooser-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <file-chooser-panel-directive></file-chooser-panel-directive>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <settings-panel-component></settings-panel-component>
        </div>

    </div>

    <div class="row bottom-row">

        <div class="col l3 m6 s12 col-backdrop">
            <detail-panel-component></detail-panel-component>
        </div>

    </div>

    <div id="titlebox">
        <a class="mwlink" href="https://www.maibornwolff.de" target="_blank"><img src="./logo.png" alt="MaibornWolff"></a>
        <h1 id="title"><span>CodeCharta {{$ctrl.pkg.version}}</span></h1>
    </div>

</div>

<code-map-component></code-map-component>