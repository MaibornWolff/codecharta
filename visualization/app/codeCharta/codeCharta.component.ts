import {UrlExtractor} from "./util/urlExtractor"
import {IHttpService, ILocationService, IRootScopeService, ITimeoutService} from "angular"
import "./codeCharta.component.scss"
import {CodeChartaService} from "./codeCharta.service"
import {SettingsService, SettingsServiceSubscriber} from "./state/settings.service";
import {ScenarioHelper} from "./util/scenarioHelper";
import {DialogService} from "./ui/dialog/dialog.service";
import {ThreeOrbitControlsService} from "./ui/codeMap/threeViewer/threeOrbitControlsService";
import {CodeMapActionsService} from "./ui/codeMap/codeMap.actions.service";
import {NameDataPair, RecursivePartial, Settings} from "./codeCharta.model"
import {FileStateService} from "./state/fileState.service";


export interface CodeChartaControllerSubscriber {
	onLoadingStatusChanged(isLoadingFile: boolean, event: angular.IAngularEvent)
}

export class CodeChartaController implements SettingsServiceSubscriber, CodeChartaControllerSubscriber {

	public static readonly LOADING_STATUS_EVENT = "loading-status-changed"

	private _viewModel: {
		version: string,
		isLoadingFile: boolean,
		isLoadingMap: boolean,
		focusedNodePath: string
	} = {
		version: require("../../package.json").version,
		isLoadingFile: true,
		isLoadingMap: true,
		focusedNodePath: ""
	}

	private urlUtils: UrlExtractor

	/* @ngInject */
	constructor(
		private threeOrbitControlsService: ThreeOrbitControlsService,
		private $rootScope: IRootScopeService,
		private dialogService: DialogService,
		private codeMapActionsService: CodeMapActionsService,
		private settingsService: SettingsService,
		private codeChartaService: CodeChartaService,
		private fileStateService: FileStateService,
		private $location: ILocationService,
		private $http: IHttpService,
		private $timeout: ITimeoutService
	) {
		SettingsService.subscribe(this.$rootScope, this)
		CodeChartaController.subscribe(this.$rootScope, this)

		this.urlUtils = new UrlExtractor(this.$location, this.$http)
		this.onLoadingStatusChanged(true, undefined)
		this.loadFileOrSample()
	}

	public onSettingsChanged(settings: Settings, update: RecursivePartial<Settings>, event: angular.IAngularEvent) {
		this._viewModel.focusedNodePath = settings.dynamicSettings.focusedNodePath
	}

	public onLoadingStatusChanged(isLoadingFile: boolean, event: angular.IAngularEvent) {
		this._viewModel.isLoadingFile = isLoadingFile
		this.synchronizeAngularTwoWayBinding()
	}

	public fitMapToView() {
		this.threeOrbitControlsService.autoFitTo()
	}

	public removeFocusedNode() {
		this.codeMapActionsService.removeFocusedNode()
	}

	public loadFileOrSample() {
		return this.urlUtils.getFileDataFromQueryParam()
			.then((data: NameDataPair[]) => {
				if (data.length > 0) {
					this.tryLoadingFiles(data)
					this.setRenderStateFromUrl();
				} else {
					this.tryLoadingSampleFiles()
				}
			})
			.catch(
				() => {
					this.tryLoadingSampleFiles()
				}
			)
	}

	public tryLoadingSampleFiles() {
		if (this.urlUtils.getParameterByName("file")) {
			this.dialogService.showErrorDialog(
				"One or more files from the given file URL parameter could not be loaded. Loading sample files instead."
			)
		}
		this.tryLoadingFiles([
            { fileName: "sample1.json", content: require("./assets/sample1.json") },
            { fileName: "sample2.json", content: require("./assets/sample2.json") }
        ]);
    }
    
    private tryLoadingFiles(values: NameDataPair[]) {
		this.settingsService.updateSettings(this.settingsService.getDefaultSettings())

		this.codeChartaService.loadFiles(values)
			.then(() => {
				this.settingsService.updateSettings(ScenarioHelper.getDefaultScenario().settings)
			})
			.catch(e => {
				this.onLoadingStatusChanged(false, undefined)
				console.error(e);
				this.printErrors(e)
			})
    }

    private setRenderStateFromUrl() {
		const renderState: string = this.urlUtils.getParameterByName("mode")
		const files = this.fileStateService.getCCFiles()

		if (renderState === "Delta" && files.length >= 2) {
			this.fileStateService.setDelta(files[0], files[1])

		} else if (renderState === "Multiple") {
			this.fileStateService.setMultiple(files)

		} else {
			this.fileStateService.setSingle(files[0])
		}
	}

	private printErrors(errors: Object) {
		this.dialogService.showErrorDialog(JSON.stringify(errors, null, "\t"))
	}

	private synchronizeAngularTwoWayBinding() {
		this.$timeout(() => {})
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: CodeChartaControllerSubscriber) {
		$rootScope.$on(CodeChartaController.LOADING_STATUS_EVENT, (event, data) => {
			subscriber.onLoadingStatusChanged(data, event)
		})
	}
}

export const codeChartaComponent = {
	selector: "codeChartaComponent",
	template: require("./codeCharta.component.html"),
	controller: CodeChartaController
}
