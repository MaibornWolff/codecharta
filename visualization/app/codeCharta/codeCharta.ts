"use strict";

import "./core/core.module";
import "./ui/ui";

import {codeChartaComponent, CodeChartaController} from "./codeCharta.component";

import angular from "angular";

angular.module(
    "app.codeCharta",
    ["app.codeCharta.core", "app.codeCharta.ui"]
);

angular.module("app.codeCharta").component(
    codeChartaComponent.selector,
    codeChartaComponent
);


