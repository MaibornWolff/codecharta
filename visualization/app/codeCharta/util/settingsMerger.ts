import {
	AttributeType,
	Edge,
	BlacklistItem,
	CCFile,
	FileSettings
} from "../codeCharta.model"
import {CodeChartaService} from "../codeCharta.service";

export class SettingsMerger {

	private static edges: Edge[] = []
	private static blacklist: BlacklistItem[] = []
	private static attributeTypesEdge: { [key: string]: AttributeType } = {}
	private static attributeTypesNode: { [key: string]: AttributeType } = {}

	public static getMergedFileSettings(inputFiles: CCFile[], withUpdatedPath: boolean = false): FileSettings {
		if (inputFiles.length == 1) {
			return inputFiles[0].settings.fileSettings ? inputFiles[0].settings.fileSettings : null
		}

		this.resetVariables()

		for (let inputFile of inputFiles) {
			this.setEdges(inputFile, withUpdatedPath)
			this.setAttributeTypesByUniqueKey(inputFile)
			this.setBlacklist(inputFile, withUpdatedPath)
			// TODO: Merge markedPackages
		}
		return this.getNewFileSettings()
	}

	private static setEdges(inputFile: CCFile, withUpdatedPath: boolean) {
		if (!inputFile.settings.fileSettings.edges) {
			return
		}

		for (let oldEdge of inputFile.settings.fileSettings.edges as Edge[]) {
			let edge: Edge = {
				fromNodeName: withUpdatedPath ? this.getUpdatedPath(inputFile.fileMeta.fileName, oldEdge.fromNodeName) : oldEdge.fromNodeName,
				toNodeName: withUpdatedPath ? this.getUpdatedPath(inputFile.fileMeta.fileName, oldEdge.toNodeName) : oldEdge.toNodeName,
				attributes: oldEdge.attributes,
				visible: oldEdge.visible
			}
			const equalEdgeItem = this.edges.find(e => e.fromNodeName == edge.fromNodeName && e.toNodeName == edge.toNodeName)

			if (equalEdgeItem) {
				for(let key in edge.attributes) {
					equalEdgeItem.attributes[key] = edge.attributes[key]
				}
			} else {
				this.edges.push(edge)
			}
		}
	}

	private static setBlacklist(inputFile: CCFile, withUpdatedPath: boolean) {
		if (!inputFile.settings.fileSettings.blacklist) {
			return
		}

		for (let oldBlacklistItem of inputFile.settings.fileSettings.blacklist as BlacklistItem[]) {
			let blacklistItem: BlacklistItem = {
				path: withUpdatedPath ? this.getUpdatedBlacklistItemPath(inputFile.fileMeta.fileName, oldBlacklistItem.path) : oldBlacklistItem.path,
				type: oldBlacklistItem.type
			}
			const equalBlacklistItems = this.blacklist.filter(b =>
				b.path == blacklistItem.path &&
				b.type == blacklistItem.type)
			if (equalBlacklistItems.length == 0) {
				this.blacklist.push(blacklistItem)
			}
		}
	}

	private static getUpdatedBlacklistItemPath(fileName: string, path: string): string {
		if (this.isAbsoluteRootPath(path)) {
			return this.getUpdatedPath(fileName, path)
		}
		return path
	}

	private static isAbsoluteRootPath(path: string): boolean {
		return path.startsWith(CodeChartaService.ROOT_PATH + "/")
	}

	private static setAttributeTypesByUniqueKey(inputFile: CCFile) {
		const types = inputFile.settings.fileSettings.attributeTypes
		if (types && types.nodes) {
			for (let key in types.nodes) {
				this.attributeTypesNode[key] = types.nodes[key]
			}
		}
		if (types && types.edges) {
			for (let key in types.edges) {
				this.attributeTypesEdge[key] = types.edges[key]
			}
		}
	}

	private static getNewFileSettings(): FileSettings {
		return {
			edges: this.edges,
			blacklist: this.blacklist,
			attributeTypes: {
				nodes: this.attributeTypesNode,
				edges: this.attributeTypesEdge
			},
			markedPackages: []
		}
	}

	private static getUpdatedPath(fileName: string, path: string): string {
		const folderArray = path.split("/")
		folderArray.splice(2, 0, fileName)
		return folderArray.join("/")
	}

	private static resetVariables() {
		this.edges = []
		this.blacklist = []
		this.attributeTypesEdge = {}
		this.attributeTypesNode = {}
	}
}
