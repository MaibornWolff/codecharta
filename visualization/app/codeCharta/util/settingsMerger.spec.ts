import {AttributeType, BlacklistItem, BlacklistType, CCFile, Edge, FileSettings, Settings} from "../codeCharta.model";
import {SettingsMerger} from "./settingsMerger";

describe("SettingsMerger", () => {

    const file1: CCFile = {
        fileMeta: {
            fileName: "file1",
            projectName: "Sample Project",
            apiVersion: "1.1"
        },
        map: {
            name: "root",
            type: "Folder",
            attributes: {},
            children: []
        },
        settings: {
            fileSettings: {
                edges: [],
                blacklist: [],
                markedPackages: [],
                attributeTypes: {}
            }
        }
    };

    const file2: CCFile = {
        fileMeta: {
            fileName: "file2",
            projectName: "Sample Project",
            apiVersion: "1.1"
        },
        map: {
            name: "root",
            type: "Folder",
            attributes: {},
            children: []
        },
        settings: {
            fileSettings: {
                edges: [],
                blacklist: [],
                markedPackages: [],
                attributeTypes: {}
            }
        }
    };

    describe("Edges merge", () => {

        let edge1: Edge
        let edge2: Edge
        let edge3: Edge
        let edge4: Edge

        beforeEach(() => {
            edge1 = {
                fromNodeName: "/root/nodeA",
                toNodeName: "/root/nodeB",
                attributes: {
                    attribute1: 10,
                    attribute2: 20
                }
            }

            edge2 = {
                fromNodeName: "/root/nodeA",
                toNodeName: "/root/nodeC",
                attributes: {
                    attribute1: 10,
                    attribute2: 20
                }
            }

            edge3 = {
                fromNodeName: "/root/nodeA",
                toNodeName: "/root/nodeB",
                attributes: {
                    attribute3: 30,
                    attribute4: 40
                }
            }

            edge4 = {
                fromNodeName: "/root/nodeA",
                toNodeName: "/root/nodeB",
                attributes: {
                    attribute1: 70,
                    attribute2: 80
                }
            }
        })


        it("should merge empty edges-arrays", () => {
            file1.settings.fileSettings.edges = [];
            file2.settings.fileSettings.edges = [];
            let fileSettings: FileSettings = SettingsMerger.getMergedFileSettings([file1, file2]);
            expect(fileSettings.edges).toEqual([]);
        });

        it("should merge different edges", () => {
            file1.settings.fileSettings.edges = [edge1];
            file2.settings.fileSettings.edges = [edge2];
            let fileSettings: FileSettings = SettingsMerger.getMergedFileSettings([file1, file2]);
            expect(fileSettings.edges).toMatchSnapshot();
        });

        it("should merge all edges if one file does not contain edges", () => {
            file1.settings.fileSettings.edges = [edge1, edge2];
            file2.settings.fileSettings.edges = null;
            let fileSettings: FileSettings = SettingsMerger.getMergedFileSettings([file1, file2]);
            expect(fileSettings.edges).toMatchSnapshot();
        });

        it("should merge edge-attributes for the same edge paths", () => {
            file1.settings.fileSettings.edges = [edge1];
            file2.settings.fileSettings.edges = [edge3];
            let fileSettings: FileSettings = SettingsMerger.getMergedFileSettings([file1, file2]);
            expect(fileSettings.edges).toMatchSnapshot();
        });

        it("should overwrite duplicated edge-attributes for the same edge", () => {
            file1.settings.fileSettings.edges = [edge1];
            file2.settings.fileSettings.edges = [edge4];
            let fileSettings: FileSettings = SettingsMerger.getMergedFileSettings([file1, file2]);
            expect(fileSettings.edges).toMatchSnapshot();
        });

    });


    describe("AttributeTypes merge", ()=> {

        const attributes1 = {
            nodes: {
                ["attribute1"]: AttributeType.absolute
            },
            edges: {
                ["attribute2"]: AttributeType.relative
            }
        };

        const attributes2 = {
            nodes: {
                ["attribute3"]: AttributeType.absolute
            },
            edges: {
                ["attribute4"]: AttributeType.relative
            }
        };

        const attributes3 = {
            nodes: {
                ["attribute1"]: AttributeType.relative
            },
            edges: {}
        };

        it("should merge different attributeTypes", () => {
            file1.settings.fileSettings.attributeTypes = attributes1;
            file2.settings.fileSettings.attributeTypes = attributes2;
            let fileSettings: FileSettings = SettingsMerger.getMergedFileSettings([file1,file2]);
            expect(fileSettings.attributeTypes).toMatchSnapshot();
        });

        it("should merge attributeTypes if one file does not contain attributeTypes", () => {
            file1.settings.fileSettings.attributeTypes = attributes1;
            file2.settings.fileSettings.attributeTypes = {};
            let fileSettings: FileSettings = SettingsMerger.getMergedFileSettings([file1,file2]);
            expect(fileSettings.attributeTypes).toMatchSnapshot();
        });

        it("should overwrite attributeType if the same exists", () => {
            file1.settings.fileSettings.attributeTypes = attributes1;
            file2.settings.fileSettings.attributeTypes = attributes3;
            let fileSettings: FileSettings = SettingsMerger.getMergedFileSettings([file1,file2]);
            expect(fileSettings.attributeTypes).toMatchSnapshot();
        });

    });

    describe("Blacklist merge", ()=> {

        const blacklistItem1: BlacklistItem = {path: "/root/nodeA", type: BlacklistType.exclude}
        const blacklistItem2: BlacklistItem = {path: "/another/nodeB", type: BlacklistType.hide}
        const blacklistItem3: BlacklistItem = {path: "/another/nodeC", type: BlacklistType.exclude}
        const blacklistItem4: BlacklistItem = {path: "*prefix/nodeD", type: BlacklistType.hide}
        const blacklistItem1Duplicate: BlacklistItem = {path: "/root/nodeA", type: BlacklistType.exclude}


        it("should merge blacklist for different paths", () => {
            file1.settings.fileSettings.blacklist = [blacklistItem1, blacklistItem2];
            file2.settings.fileSettings.blacklist = [blacklistItem3, blacklistItem4];
            let fileSettings = SettingsMerger.getMergedFileSettings([file1,file2]);
            expect(fileSettings.blacklist).toMatchSnapshot();
        });

        it("should only contain unique paths+type", () => {
            file1.settings.fileSettings.blacklist = [blacklistItem1, blacklistItem2];
            file2.settings.fileSettings.blacklist = [blacklistItem1Duplicate, blacklistItem4];
            let fileSettings = SettingsMerger.getMergedFileSettings([file1,file2]);
            expect(fileSettings.blacklist).toMatchSnapshot();
        });
    });
});