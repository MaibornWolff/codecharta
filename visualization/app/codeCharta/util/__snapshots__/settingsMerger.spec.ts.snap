// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsMerger AttributeTypes merge should merge attributeTypes if one file does not contain attributeTypes 1`] = `
Object {
  "edges": Object {
    "attribute2": 1,
  },
  "nodes": Object {
    "attribute1": 0,
  },
}
`;

exports[`SettingsMerger AttributeTypes merge should merge different attributeTypes 1`] = `
Object {
  "edges": Object {
    "attribute2": 1,
    "attribute4": 1,
  },
  "nodes": Object {
    "attribute1": 0,
    "attribute3": 0,
  },
}
`;

exports[`SettingsMerger AttributeTypes merge should overwrite attributeType if the same exists 1`] = `
Object {
  "edges": Object {
    "attribute2": 1,
  },
  "nodes": Object {
    "attribute1": 1,
  },
}
`;

exports[`SettingsMerger Blacklist merge should merge blacklist for different paths 1`] = `
Array [
  Object {
    "path": "/root/nodeA",
    "type": "exclude",
  },
  Object {
    "path": "/another/nodeB",
    "type": "hide",
  },
  Object {
    "path": "/another/nodeC",
    "type": "exclude",
  },
  Object {
    "path": "*prefix/nodeD",
    "type": "hide",
  },
]
`;

exports[`SettingsMerger Blacklist merge should only contain unique paths+type 1`] = `
Array [
  Object {
    "path": "/root/nodeA",
    "type": "exclude",
  },
  Object {
    "path": "/another/nodeB",
    "type": "hide",
  },
  Object {
    "path": "*prefix/nodeD",
    "type": "hide",
  },
]
`;

exports[`SettingsMerger Edges merge should merge all edges if one file does not contain edges 1`] = `
Array [
  Object {
    "attributes": Object {
      "attribute1": 10,
      "attribute2": 20,
    },
    "fromNodeName": "/root/nodeA",
    "toNodeName": "/root/nodeB",
    "visible": undefined,
  },
  Object {
    "attributes": Object {
      "attribute1": 10,
      "attribute2": 20,
    },
    "fromNodeName": "/root/nodeA",
    "toNodeName": "/root/nodeC",
    "visible": undefined,
  },
]
`;

exports[`SettingsMerger Edges merge should merge different edges 1`] = `
Array [
  Object {
    "attributes": Object {
      "attribute1": 10,
      "attribute2": 20,
    },
    "fromNodeName": "/root/nodeA",
    "toNodeName": "/root/nodeB",
    "visible": undefined,
  },
  Object {
    "attributes": Object {
      "attribute1": 10,
      "attribute2": 20,
    },
    "fromNodeName": "/root/nodeA",
    "toNodeName": "/root/nodeC",
    "visible": undefined,
  },
]
`;

exports[`SettingsMerger Edges merge should merge edge-attributes for the same edge paths 1`] = `
Array [
  Object {
    "attributes": Object {
      "attribute1": 10,
      "attribute2": 20,
      "attribute3": 30,
      "attribute4": 40,
    },
    "fromNodeName": "/root/nodeA",
    "toNodeName": "/root/nodeB",
    "visible": undefined,
  },
]
`;

exports[`SettingsMerger Edges merge should overwrite duplicated edge-attributes for the same edge 1`] = `
Array [
  Object {
    "attributes": Object {
      "attribute1": 70,
      "attribute2": 80,
    },
    "fromNodeName": "/root/nodeA",
    "toNodeName": "/root/nodeB",
    "visible": undefined,
  },
]
`;
